pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go

```diff
@@ -135,34 +142,45 @@ func RedirectStep(cache *Cache, stepNum int, step *corev1.Container, kubeclient
 	step.Args = GetArgs(stepNum, step.Command, step.Args)
 	step.Command = []string{BinaryLocation}
 	step.VolumeMounts = append(step.VolumeMounts, toolsMount)
+	if stepNum == 0 {
+		step.VolumeMounts = append(step.VolumeMounts, downwardMount)
+	}
 	return nil
 }
 
 // GetArgs returns the arguments that should be specified for the step which has been wrapped
 // such that it will execute our custom entrypoint instead of the user provided Command and Args.
 func GetArgs(stepNum int, commands, args []string) []string {
-	waitFile := fmt.Sprintf("%s/%s", MountPoint, strconv.Itoa(stepNum-1))
-	if stepNum == 0 {
-		waitFile = ""
-	}
+	waitFile := getWaitFile(stepNum)
 	// The binary we want to run must be separated from its arguments by --
 	// so if commands has more than one value, we'll move the other values
 	// into the arg list so we can separate them
 	if len(commands) > 1 {
 		args = append(commands[1:], args...)
 		commands = commands[:1]
 	}
-	argsForEntrypoint := append([]string{
+	argsForEntrypoint := []string{
 		"-wait_file", waitFile,
-		"-post_file", fmt.Sprintf("%s/%s", MountPoint, strconv.Itoa(stepNum)),
-		"-entrypoint"},
-		commands...,
-	)
+		"-post_file", getWaitFile(stepNum + 1),
+	}
+	if stepNum == 0 {
+		argsForEntrypoint = append(argsForEntrypoint, "-wait_file_content")
+	}
+	argsForEntrypoint = append(argsForEntrypoint, "-entrypoint")
+	argsForEntrypoint = append(argsForEntrypoint, commands...)
 	// TODO: what if Command has multiple elements, do we need "--" between command and args?
 	argsForEntrypoint = append(argsForEntrypoint, "--")
 	return append(argsForEntrypoint, args...)
 }
```

Comment:
pretty cool that you were able to clean this up - and the tests didnt even have to change :) nice improvements!